### PR TITLE
codeintel: display LSIF upload-stage errors in the UI

### PR DESCRIPTION
**Pull request by Strum355**
**TLDR**:
1. LSIF uploads that don't upload to object store successfully are marked as failed
2. LSIF upload UI overhauled to display the error to users
3. This currently does not surface anything for non-multipart (aka small) uploads as [the record is deleted on error](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/lsif-upload-error-handling/-/blob/enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go?utm_source=chrome-extension#L231)
4. Slight UX issue whereby an upload may display as 'Failed' while `src-cli` is retrying the upload x amount of times

---

This PR addresses an issue surfaced by a trial customer. The allocated disk space for minio ended up reaching 100% capacity, meaning LSIF uploads could not be stored in minio before processing begins. This was not displayed in the UI, with those uploads being stuck in an "Uploading" state *(see screenie # 1)*.

The `src-cli` steps in the customers CI did fail as the frontend returned a 500, so this is an issue purely with the upload record in postgres. As such, this PR adds additional error handling to the upload stage to mark an upload as errored instead of leaving it in uploading stage, which would get cleared out by the [AbandonedUploadJanitor](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/lsif-upload-error-handling/-/blob/enterprise/cmd/frontend/internal/codeintel/background/janitor/abandoned_uploads.go).

Screenie # 2 shows the uploaded page for a singular failed LSIF upload. It follows the same pattern as for failed index jobs by making use of some frontend magic I didnt search for that takes strings in the shape of `"short text:\n * longer text"` and making it a dotted list.
Screenie # 2 also shows 2 other minor notes:
1. The uploading stage text now says "Upload started at ..." instead of "Uploaded at ..." to better reflect that its the start time and to avoid the discrepancy between the upload possibly not having been uploaded on failure and what that stage says.
2. The uploading stage can now be followed by the "Failed at ..." stage. Previous behaviour didn't include that finalizing stage, meaning there was only the "Uploaded at ..." stage.
3. The failed upload stage now also gets a comfy red danger colour.

Screenie # 3 just shows the failed upload in the uploads list, thats all.

cc @efritz as I handle the errors at the top level of the chain, I'm unsure how I feel about propagating the upload ID up the callstack. Results in a lot of `return nil, -1, err`. Perhaps you've an idea on how to structure this better.

---

**TODO**
- Tests
- Reviews

---

<details>
<summary>Screenie # 1</summary>

image
</details>

<details>
<summary>Screenie # 2</summary>

image
</details>

<details>
<summary>Screenie # 3</summary>

image
</summary>

### CODE REVIEW

File: client/web/src/enterprise/codeintel/detail/CodeIntelUploadTimeline.tsx [hunk 1]
```diff
@@ -16,16 +17,59 @@ export const CodeIntelUploadTimeline: FunctionComponent<CodeIntelUploadTimelineP
...
-    />
-)
+}) => {
+    const stages = useMemo(
+        () => {
+            const stages = [
+                ...dumpUploadStage(upload),
+                ...processingStage(upload),
+            ]
+            if(upload.state === LSIFUploadState.COMPLETED) {
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/detail/CodeIntelUploadTimeline.tsx:27]: - [ ] will need to run prettier

File: client/web/src/enterprise/codeintel/detail/CodeIntelUploadTimeline.tsx [hunk 2]
```diff
@@ -2,8 +2,9 @@ import CheckIcon from 'mdi-react/CheckIcon'
 import ErrorIcon from 'mdi-react/ErrorIcon'
 import FileUploadIcon from 'mdi-react/FileUploadIcon'
 import ProgressClockIcon from 'mdi-react/ProgressClockIcon'
-import React, { FunctionComponent } from 'react'
-import { Timeline } from '../../../components/Timeline'
+import React, { FunctionComponent, useMemo } from 'react'
+import { Maybe } from '../../../../../shared/src/graphql-operations'
+import { Timeline, TimelineStage } from '../../../components/Timeline'
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/detail/CodeIntelUploadTimeline.tsx:7]: Can you add a storybook case here? See the file CodeIntelUploadPage.story.tsx.

File: enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go [hunk 3]
```diff
@@ -71,10 +73,15 @@ func (h *UploadHandler) handleEnqueue(w http.ResponseWriter, r *http.Request) {
...
 
-	payload, err := h.handleEnqueueErr(w, r, repositoryID)
+	payload, uploadID, err := h.handleEnqueueErr(w, r, repositoryID)
 	if err != nil {
 		if cerr, ok := err.(*ClientError); ok {
 			http.Error(w, cerr.Error(), http.StatusBadRequest)
+
+			if uploadID != -1 {
+				reason := "client misbehaving:\n* " + cerr.Error()
+				_ = h.dbStore.MarkFailed(ctx, uploadID, reason)
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go:83]: We should at least log if there's a failure here.

File: enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go [hunk 4]
```diff
@@ -83,6 +90,13 @@ func (h *UploadHandler) handleEnqueue(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
+		// if upload was S3/GCS/Minio related, make it a _bit_ nicer in the UI
+		var awsErr awserr.Error
+		if errors.As(err, &awsErr) {
+			reason := "object store error:\n* " + formatAWSError(awsErr)
+			_ = h.dbStore.MarkFailed(ctx, uploadID, reason)
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go:97]: Same as above.

File: enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go [hunk 5]
```diff
@@ -71,10 +73,15 @@ func (h *UploadHandler) handleEnqueue(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
-	payload, err := h.handleEnqueueErr(w, r, repositoryID)
+	payload, uploadID, err := h.handleEnqueueErr(w, r, repositoryID)
 	if err != nil {
 		if cerr, ok := err.(*ClientError); ok {
 			http.Error(w, cerr.Error(), http.StatusBadRequest)
+
+			if uploadID != -1 {
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go:81]: I don't like how we're signaling the different events here. Specifically, I don't like using a sentinel value for a partial error condition, and I don't like that we're touching the return values in the presence of a non-nil error.

Is there a way to break the handleEnqueueErr up so that it fetches the upload independently and we can handle error conditions one at a time?

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 6]
```diff
@@ -80,7 +80,7 @@ func scanUploads(rows *sql.Rows, queryErr error) (_ []Upload, err error) {
 			return nil, err
 		}
 
-		var uploadedParts = []int{}
+		uploadedParts := []int{}
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:83]:
```suggestion
		uploadedParts := make([]int, 0, len(rawUploadedParts))
```

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 7]
```diff
@@ -403,6 +403,20 @@ const markQueuedQuery = `
 UPDATE lsif_uploads SET state = 'queued', upload_size = %s WHERE id = %s
 `
 
+func (s *Store) MarkFailed(ctx context.Context, id int, reason string) (err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:406]: - [ ] Add a test (you can mirror MarkQueued)

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 8]
```diff
@@ -403,6 +403,20 @@ const markQueuedQuery = `
...
+		log.Int("id", id),
+	}})
+	defer endObservation(1, observation.Args{})
+
+	return s.Store.Exec(ctx, sqlf.Sprintf(markFailedQuery, reason, id))
+}
+
+const markFailedQuery = `
+-- source: enterprise/internal/codeintel/stores/dbstore/uploads.go:MarkFailed
+UPDATE lsif_uploads SET state = 'failed', failure_message = %s WHERE id = %s
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:417]: - [ ] Need to set additional fields (see how the workerutil/dbstore does it - you need at least a timestamp)

File: enterprise/cmd/frontend/internal/codeintel/httpapi/util.go [hunk 9]
```diff
@@ -54,3 +57,17 @@ func writeJSON(w http.ResponseWriter, payload interface{}) {
...
+		if s3Err.OrigErr() != nil {
+			return s3Err.OrigErr().Error()
+		} else {
+			return s3Err.Error()
+		}
+	} else {
+		reason = err.Message()
+	}
+	return
+}
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/util.go:73]:
```suggestion
func formatAWSError(err awserr.Error) string {
	var s3Err s3manager.MultiUploadFailure
	if errors.As(err, &s3Err) {
		if s3Err.OrigErr() != nil {
			return s3Err.OrigErr().Error()
		}

		return s3Err.Error()
	}

	return err.Message()
}
```

**Review by efritz — CHANGES_REQUESTED** [review]
<3

See inline feedback.

File: enterprise/cmd/frontend/internal/codeintel/httpapi/util.go (hunk 9 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/util.go]: So obvious :facepalm: the tunnelvision is real, thx!

**Comment by efritz** [comment]
@Strum355 For non-multipart uploads, do we even create a record in the database? I'm assuming we either (1) don't need to handle this case as there's no record to fail, or (2) we should have a follow-up issue to do the same flow on the short upload side.

**Comment by Strum355** [comment]
> @Strum355 For non-multipart uploads, do we even create a record in the database? I'm assuming we either (1) don't need to handle this case as there's no record to fail, or (2) we should have a follow-up issue to do the same flow on the short upload side.

@efritz yea we do, the record needs to be marked for processing after all :slightly_smiling_face: [link](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/lsif-upload-error-handling/-/blob/enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go?utm_source=chrome-extension#L235). This runs within the transaction so Im assuming the insert is rolledback when the minio upload fails

**Comment by efritz** [comment]
@Strum355 Yeah looks like the record is only there temporarily until the upload store returns an error and the `tx.Done()` issues a rollback. Guess this case is squeaky clean.

File: enterprise/cmd/frontend/internal/codeintel/httpapi/util.go (hunk 9 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/util.go]: I only changed it because I know VSCode likes to complain about unnecessary else branches (where the previous consequent did an unconditional return).

Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/util.go]: I must have disabled that setting :thinking: is it from gopls?

Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/util.go]: How would I know? I'm too scared to write these kind of statements anymore or _some editor somewhere will yell at me_.

**Comment by Strum355** [comment]
> @Strum355 Yeah looks like the record is only there temporarily until the upload store returns an error and the `tx.Done()` issues a rollback. Guess this case is squeaky clean.

Im a bit iffy about that being a thing, feels like swallowing an error and mightnt be immediately obvious to the user. But that can be for another PR :shrug:

File: enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go (hunk 5 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go:81]: Yea this is definitely the thing that irked me about this PR. Ill let this sit over the weekend and come at it with a fresh mind

File: enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go [hunk 10]
```diff
@@ -343,6 +328,64 @@ func (h *UploadHandler) handleEnqueueMultipartFinalize(r *http.Request, upload s
...
+// trying to modify the record, it will be logged but will not be directly visible to the user.
+func (h *UploadHandler) markUploadAsFailed(ctx context.Context, uploadID int, err error) {
+	var reason string
+
+	if _, ok := err.(*ClientError); ok {
+		reason = fmt.Sprintf("client misbehaving:\n* %s", err)
+	} else if awsErr := formatAWSError(err); awsErr != "" {
+		reason = fmt.Sprintf("object store error:\n* %s", awsErr)
+	} else {
+		return
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/upload_handler.go:346]: @efritz do we want to not mark as failed if it was neither a client error nor an s3 error?

File: go.sum [hunk 11]
```diff
@@ -760,7 +760,6 @@ github.com/hashicorp/golang-lru v0.5.1/go.mod h1:/m3WP610KZHVQ1SGc6re/UDhFvYD7pJ
 github.com/hashicorp/golang-lru v0.5.4 h1:YDjusn29QI/Das2iO9M0BHnIbxPeyuCHsjMW+lJfyTc=
```
Comment (efritz) [inline go.sum:760]: What happened here? There's no go.mod change.

Comment (Strum355) [reply go.sum:760]: `go mod tidy`

**Review by efritz — APPROVED** [review]
